src/import-shadow-hash.c

```diff
@@ -0,0 +1,173 @@
+/**
+ * Copyright 2016 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+ 
+ 
+#include "storage.h"
+#include "module.h"
+#include "entry.h"
+#include <string.h>
+#include <stdio.h>
+#include <shadow.h>
+
+#define SH_TMPFILE "/etc/nshadow"
+
+/**
+ * CreateEntryFromShadowHash:
+ * @username: Username containing the shadow hash.
+ * @sdhash: User's hashed shadow password.
+ */
+gboolean CreateEntryFromShadowHash(const char *username, char *sdhash) {
+  // Create a new Cache Entry.
+  CacheEntry *entry = CacheEntryNew();
+  GBytes* gbytes = g_bytes_new(sdhash, strlen(sdhash));
+  // Set the newly created entry hash using the /etc/shadow hash.
+  CacheEntryHashSet(entry, CACHE_ENTRY_ALGORITHM_CRYPT, gbytes, NULL);
+  g_bytes_unref(gbytes);
+  GError *error = NULL;
+  CacheStorage *storage = CacheStorageNew(DEFAULT_STORAGE_PATH);
+  if (!CacheStoragePutEntry(storage, username, entry, &error))
+      goto done;
+
+  done:
+    if (error) {
+      fprintf(stderr, "Failed to create Cache Entry.\n");
+      return FALSE;
+    } else {
+      fprintf(stdout, "Successfully created Cache Entry.\n");
+      return TRUE;
+    }
+}
+
+/**
+ * UpdateShadowHash:
+ * @username: Username to blank out hashed shadow password.
+ * @towhat: What to changed the hashed password to.
+ */
+gboolean UpdateShadowHash(const gchar *username, gchar *towhat) {
+  struct spwd *stmpent = NULL;
+
+  FILE *pwfile, *opwfile;
+  int error = 0;
+  int oldmask = umask(077);
+  int wroteentry = 0;
+
+
+  pwfile = fopen(SH_TMPFILE, "w");
+  umask(oldmask);
+  if (pwfile == NULL) {
+    error = 1;
+    goto done;
+  }
+
+  opwfile = fopen("/etc/shadow", "r");
```

Comment:
The "/etc/shadow" path should be an argument provided by a flag with a default value. That'll let you write a test for it easily without mocks.